Gemfile

```diff
@@ -14,3 +14,5 @@ gem 'rubocop', '~> 1.21'
 gem 'mediawiki_api'
 
 gem 'json'
+
+gem 'deep_merge'
```

Comment:
Hmm... In the Dashboard codebase, we use the deep_merge functions (`Hash#deep_merge!` specifically) that are built into Ruby on Rails via ActiveSupport: https://apidock.com/rails/ActiveSupport/DeepMergeable/deep_merge%21

I think I would prefer adding that as a dependency rather than the separate deep_merge gem, so that this change won't expand the Dashboard's dependencies.

Reply:
To do this, I added `require 'active_support/core_ext/hash/deep_merge'`

But when i ran it, using the existing `data.deep_merge!(response.data)`, it processed only 17 instead of 25, which is the same behaviour as just using `.merge`, this time not because of missing revisions but rather, the ActiveSupport deep_merge function  doesn't handle the response.data properly for some reason, it probably overwrites the affected revisions because of duplicate keys or something.

In the documentation link you sent, it did say that 'A block can be provided to merge non-DeepMergeable values' so using the response json structure for reference, I added the block below and it works, and processes all 25 revisions.

```
data.deep_merge!(response.data) do |key, old_val, new_val|
      if key == 'revisions' && old_val.is_a?(Array) && new_val.is_a?(Array)
        old_val + new_val  
      else
        new_val  
      end
    end
```
Comparison: (The ones denoted in red, are the ones not processed without the block
![image](https://github.com/user-attachments/assets/29d418c1-74bb-4a82-abe6-17ae1c1f7264)

Reply:
> So should I use this block?

Maybe. I'd like to get an explanation of why exactly that version of deep_merge! behaves differently first.

Reply:
From my understanding + checking docs for the 2 versions: [active_support deep merge](https://www.rubydoc.info/github/rails/rails/ActiveSupport/DeepMergeable#deep_merge!-instance_method) vs [deep_merge! gem](https://www.rubydoc.info/gems/deep_merge/DeepMerge#deep_merge!-class_method)

- For the gem, the docs say:
"Deep Merge core documentation. deep_merge! method permits merging of arbitrary child elements. The two top level elements must be hashes. These hashes can contain unlimited (to stack limit) levels of child elements. These child elements to not have to be of the same types. Where child elements are of the same type, deep_merge will attempt to merge them together. " 
So as long as the top level elements are hashes , it will recursively merge any nested child elements within the hashes, regardless of how deeply nested they are. so it just merges all the hashes in the revisions array.

- The Active Support 'deep_merge!' method is optimized for merging hashes and not necessarily for handling complex nested arrays in a way that preserves all elements. If the data structure includes arrays (like revisions), the merging logic might inadvertently drop or skip some array elements.
It could be that Active Support’s deep_merge! is treating the affected revisions as conflicts that it cannot reconcile.
Or the deep_merge! method is encountering the revisions array at a level where it expects to merge hashes, so it chooses to skip those revisions instead of merging them.

But as to why those specific revisions are skipped /overwritten using the ActiveSupport method, unfortunately I cannot tell, I saved the incomplete / complete responses to analyze it, but the hashes for the missing revisions ids are just not present in the revisions array, there isn't a broken structure or anything.

example response structure:
```
"95116959": {
      "pageid": 95116959,
      "ns": 0,
      "title": "Q96417649",
      "revisions": [
        {
          "revid": 2266123060,
          "parentid": 2265005040,
          "slots": {
            "main": {
              "contentmodel": "wikibase-item",
              "contentformat": "application/json",
              "*": "{//insert content here}"
            }
          },
          "comment": "/* wbsetdescription-set:1|jv */ gim vidéo, #quickstatements; #temporary_batch_1730083092317"
        },
        {
          "revid": 2266123123,
          "parentid": 2266123060,
          "slots": {
            "main": {
              "contentmodel": "wikibase-item",
              "contentformat": "application/json",
              "*": "{//insert content here}"
            }
          },
          "comment": "/* wbsetdescription-add:1|su */ kaulinan video, #quickstatements; #temporary_batch_1730083092317"
        }, ....
```